Leverage Babel when choosing the translations and displaying them

lerks opened the pull request:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cms-dev/cms/844)
<!-- Reviewable:end -->

stefano-maggiolo commented:
Review status: 0 of 7 files reviewed at latest revision, all discussions resolved.

---

*[cms/locale/locale.py, line 337 at r2](https://beta.reviewable.io/reviews/cms-dev/cms/844#-L5ei5H5BTkWD5tZQ_-T:-L5ei5H5BTkWD5tZQ_-U:b7wlxxw) ([raw file](https://github.com/cms-dev/cms/blob/1ee07e0fdecf2948722282f3ad7c9e9d3ef63714/cms/locale/locale.py#L337)):*
> ```Python
>             logger.error("Invalid installed locale: %s", lang_code)
>         else:
>             parsed_lang_codes.append(parsed_lang_code)
> ```

You can do this in the try

---

*[cms/locale/locale.py, line 346 at r2](https://beta.reviewable.io/reviews/cms-dev/cms/844#-L5eiAOu9A45Pymis9N9:-L5eiAOu9A45Pymis9NA:bb73xfv) ([raw file](https://github.com/cms-dev/cms/blob/1ee07e0fdecf2948722282f3ad7c9e9d3ef63714/cms/locale/locale.py#L346)):*
> ```Python
>             logger.error("Invalid allowed locale: %s", prefix)
>         else:
>             parsed_prefixes.append(parsed_prefix)
> ```

Same

---

*[cms/locale/locale.py, line 352 at r2](https://beta.reviewable.io/reviews/cms-dev/cms/844#-L5eiWhd-BYLgRVjP1p-:-L5eiWhd-BYLgRVjP1p0:br1mhlg) ([raw file](https://github.com/cms-dev/cms/blob/1ee07e0fdecf2948722282f3ad7c9e9d3ef63714/cms/locale/locale.py#L352)):*
> ```Python
>         for parsed_prefix in parsed_prefixes:
>             if all(p_c is None or lc_c == p_c
>                    for lc_c, p_c in zip(parsed_lang_code, parsed_prefix)):
> ```

I'm confused. Care to make more obvious what's going on? Aren't parsed_lang_code and parsed_prefix strings or something similar?

---

*[cms/server/util.py, line 417 at r1](https://beta.reviewable.io/reviews/cms-dev/cms/844#-L5dwfKf03gWN5DN6n0O:-L5dwfKf03gWN5DN6n0P:b-vq2itf) ([raw file](https://github.com/cms-dev/cms/blob/e9e2b62d830babc2e884c7a889b8a75ccbdb07e7/cms/server/util.py#L417)):*
> ```Python
>         self.timestamp = make_datetime()
>         self.sql_session = Session()
>         self.sql_session.expire_all()
> ```

Not related to the PR - is this even useful? We just created that session!

---

*[cms/server/util.py, line 418 at r1](https://beta.reviewable.io/reviews/cms-dev/cms/844#-L5dxdYb3T-FxpfnglLk:-L5dxdYb3T-FxpfnglLl:b-wo7cl7) ([raw file](https://github.com/cms-dev/cms/blob/e9e2b62d830babc2e884c7a889b8a75ccbdb07e7/cms/server/util.py#L418)):*
> ```Python
>         self.sql_session = Session()
>         self.sql_session.expire_all()
>         self.r_params = dict()
> ```

I think something relies on this being None initially, e.g. I think this breaks `write_error`.

Edit: oh, I see you tried to change later - but why? It doesn't make sense to test if this was filled by looking at the length, it might be intentionally empty...

(Unrelated: I think codacy doesn't like much testing for emptiness with len)

---

*[cms/server/contest/handlers/base.py, line 63 at r1](https://beta.reviewable.io/reviews/cms-dev/cms/844#-L5dyBuJBGeSx-V-zl7U:-L5dyBuJBGeSx-V-zl7V:b-2p3ex8) ([raw file](https://github.com/cms-dev/cms/blob/e9e2b62d830babc2e884c7a889b8a75ccbdb07e7/cms/server/contest/handlers/base.py#L63)):*
> ```Python
>         super(BaseHandler, self).__init__(*args, **kwargs)
>         self.available_translations = self.service.translations
>         self.browser_lang = "en"
> ```

Can you retrieve this from `DEFAULT_TRANSLATION`?

---

*[cms/server/contest/handlers/base.py, line 78 at r1](https://beta.reviewable.io/reviews/cms-dev/cms/844#-L5dz8bn-UV8Iv_p37_K:-L5dz8bn-UV8Iv_p37_L:b-l8ohsn) ([raw file](https://github.com/cms-dev/cms/blob/e9e2b62d830babc2e884c7a889b8a75ccbdb07e7/cms/server/contest/handlers/base.py#L78)):*
> ```Python
>     def get(self):
>         self.render_params()
> ```

Unrelated: this blank line hurts my eyes

---

*[cms/server/contest/handlers/base.py, line 85 at r1](https://beta.reviewable.io/reviews/cms-dev/cms/844#-L5dzEGG46aIJyIM4G2M:-L5dzEGG46aIJyIM4G2N:bw7duay) ([raw file](https://github.com/cms-dev/cms/blob/e9e2b62d830babc2e884c7a889b8a75ccbdb07e7/cms/server/contest/handlers/base.py#L85)):*
> ```Python
> 
>         """
>         super(BaseHandler, self).prepare()
> ```

Remove entirely?

---

*[cms/server/contest/handlers/base.py, line 92 at r1](https://beta.reviewable.io/reviews/cms-dev/cms/844#-L5dzPqs4fG933iffsO7:-L5dzPqs4fG933iffsO8:bpso3yg) ([raw file](https://github.com/cms-dev/cms/blob/e9e2b62d830babc2e884c7a889b8a75ccbdb07e7/cms/server/contest/handlers/base.py#L92)):*
> ```Python
>         """
>         self.r_params["now"] = self.timestamp
>         self.r_params["url"] = self.url
> ```

You missed contest_list

---

*[cms/server/contest/handlers/base.py, line 64 at r2](https://beta.reviewable.io/reviews/cms-dev/cms/844#-L5ejBE94AtRxNlCyEht:-L5ejBE94AtRxNlCyEhu:b-brw89v) ([raw file](https://github.com/cms-dev/cms/blob/1ee07e0fdecf2948722282f3ad7c9e9d3ef63714/cms/server/contest/handlers/base.py#L64)):*
> ```Python
>         self.available_translations = self.service.translations
>         self.automatic_translation = DEFAULT_TRANSLATION
>         self.cookie_translation = None
> ```

Comments on what *_translation represent?

---

*[cms/server/contest/handlers/contest.py, line 333 at r1](https://beta.reviewable.io/reviews/cms-dev/cms/844#-L5e--oKBWU4fSOoVFM5:-L5e--oKBWU4fSOoVFM6:bb76tpr) ([raw file](https://github.com/cms-dev/cms/blob/e9e2b62d830babc2e884c7a889b8a75ccbdb07e7/cms/server/contest/handlers/contest.py#L333)):*
> ```Python
> 
>         self.set_header("Content-Language", lang_code)
>         self.translation = self.available_translations[lang_code.replace("-", "_")]
> ```

Wrap

---

*[cms/server/contest/handlers/contest.py, line 413 at r1](https://beta.reviewable.io/reviews/cms-dev/cms/844#-L5e-6LgEbg8pKKQFokk:-L5e-6LgEbg8pKKQFokl:b-7t3792) ([raw file](https://github.com/cms-dev/cms/blob/e9e2b62d830babc2e884c7a889b8a75ccbdb07e7/cms/server/contest/handlers/contest.py#L413)):*
> ```Python
> 
>         self.r_params["translation"] = self.translation
>         self.r_params["_"] = self._
> ```

I would have preferred the change in render_params()`s return value to be in a separate commit

---

*[cms/server/contest/handlers/contest.py, line 320 at r2](https://beta.reviewable.io/reviews/cms-dev/cms/844#-L5ek1RV61waxK1XCXaL:-L5ek1RV61waxK1XCXaM:b-2ntjb) ([raw file](https://github.com/cms-dev/cms/blob/1ee07e0fdecf2948722282f3ad7c9e9d3ef63714/cms/server/contest/handlers/contest.py#L320)):*
> ```Python
>         automatic_lang = choose_language_code(browser_langs, lang_codes)
>         if automatic_lang is None:
>             automatic_lang = lang_codes[0]
> ```

What if no avail translations?

---

*[cms/server/contest/handlers/contest.py, line 326 at r2](https://beta.reviewable.io/reviews/cms-dev/cms/844#-L5ek6Fp6gQKH0YU5epA:-L5ek6Fp6gQKH0YU5epB:b-ipb8oa) ([raw file](https://github.com/cms-dev/cms/blob/1ee07e0fdecf2948722282f3ad7c9e9d3ef63714/cms/server/contest/handlers/contest.py#L326)):*
> ```Python
>         cookie_lang = self.get_cookie("language", None)
>         if cookie_lang is not None:
>             self.cookie_translation = self.available_translations[cookie_lang]
> ```

What if cookie_lang is not available?

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cms-dev/cms/844)*
<!-- Sent from Reviewable.io -->

lerks commented:
Review status: 0 of 7 files reviewed at latest revision, 14 unresolved discussions.

---

*[cms/locale/locale.py, line 337 at r2](https://beta.reviewable.io/reviews/cms-dev/cms/844#-L5ei5H5BTkWD5tZQ_-T:-L5euUz0-mtnary7JxMf:bik6ntp) ([raw file](https://github.com/cms-dev/cms/blob/1ee07e0fdecf2948722282f3ad7c9e9d3ef63714/cms/locale/locale.py#L337)):*
<details><summary><i>Previously, stefano-maggiolo (Stefano Maggiolo) wrote…</i></summary><blockquote>

You can do this in the try
</blockquote></details>

I prefer to keep in the try only the instructions that can fail, I think it makes the flows clearer and more readable. Do you feel strongly about this?

---

*[cms/locale/locale.py, line 346 at r2](https://beta.reviewable.io/reviews/cms-dev/cms/844#-L5eiAOu9A45Pymis9N9:-L5eujHi6f27clKPOEj2:bb73xfv) ([raw file](https://github.com/cms-dev/cms/blob/1ee07e0fdecf2948722282f3ad7c9e9d3ef63714/cms/locale/locale.py#L346)):*
<details><summary><i>Previously, stefano-maggiolo (Stefano Maggiolo) wrote…</i></summary><blockquote>

Same
</blockquote></details>

Same

---

*[cms/locale/locale.py, line 352 at r2](https://beta.reviewable.io/reviews/cms-dev/cms/844#-L5eiWhd-BYLgRVjP1p-:-L5euoS50YNMFztToxLQ:b-y6mluf) ([raw file](https://github.com/cms-dev/cms/blob/1ee07e0fdecf2948722282f3ad7c9e9d3ef63714/cms/locale/locale.py#L352)):*
<details><summary><i>Previously, stefano-maggiolo (Stefano Maggiolo) wrote…</i></summary><blockquote>

I'm confused. Care to make more obvious what's going on? Aren't parsed_lang_code and parsed_prefix strings or something similar?
</blockquote></details>

They are tuples.  I added a comment.

---

*[cms/server/util.py, line 417 at r1](https://beta.reviewable.io/reviews/cms-dev/cms/844#-L5dwfKf03gWN5DN6n0O:-L5ewLmd8h5Z5YDNh7Lo:b-wpm3b3) ([raw file](https://github.com/cms-dev/cms/blob/e9e2b62d830babc2e884c7a889b8a75ccbdb07e7/cms/server/util.py#L417)):*
<details><summary><i>Previously, stefano-maggiolo (Stefano Maggiolo) wrote…</i></summary><blockquote>

Not related to the PR - is this even useful? We just created that session!
</blockquote></details>

I don't know. I also think that it's useless. Though fixing that is out of scope. You can send me a commit if you want. :)

---

*[cms/server/util.py, line 418 at r1](https://beta.reviewable.io/reviews/cms-dev/cms/844#-L5dxdYb3T-FxpfnglLk:-L5exASb4m0lH7kCeCWE:b5tm1tw) ([raw file](https://github.com/cms-dev/cms/blob/e9e2b62d830babc2e884c7a889b8a75ccbdb07e7/cms/server/util.py#L418)):*
<details><summary><i>Previously, stefano-maggiolo (Stefano Maggiolo) wrote…</i></summary><blockquote>

I think something relies on this being None initially, e.g. I think this breaks `write_error`.

Edit: oh, I see you tried to change later - but why? It doesn't make sense to test if this was filled by looking at the length, it might be intentionally empty...

(Unrelated: I think codacy doesn't like much testing for emptiness with len)
</blockquote></details>

I thought it was simpler if the rest of the code could assume `r_params` was always a dict, rather than allowing it to be None. But I don't feel strongly about this, I can revert it if you prefer.

(I don't see any codacy comments, why do you think it doesn't like this?)

---

*[cms/server/contest/handlers/base.py, line 63 at r1](https://beta.reviewable.io/reviews/cms-dev/cms/844#-L5dyBuJBGeSx-V-zl7U:-L5evs1LBbIjS1nLKGro:bcchv37) ([raw file](https://github.com/cms-dev/cms/blob/e9e2b62d830babc2e884c7a889b8a75ccbdb07e7/cms/server/contest/handlers/base.py#L63)):*
<details><summary><i>Previously, stefano-maggiolo (Stefano Maggiolo) wrote…</i></summary><blockquote>

Can you retrieve this from `DEFAULT_TRANSLATION`?
</blockquote></details>

Will do before pushing, as this is overwritten later (by exactly what you are suggesting)

---

*[cms/server/contest/handlers/base.py, line 78 at r1](https://beta.reviewable.io/reviews/cms-dev/cms/844#-L5dz8bn-UV8Iv_p37_K:-L5f0J7m4UY9qw_AEb2Z:bss4a2b) ([raw file](https://github.com/cms-dev/cms/blob/e9e2b62d830babc2e884c7a889b8a75ccbdb07e7/cms/server/contest/handlers/base.py#L78)):*
<details><summary><i>Previously, stefano-maggiolo (Stefano Maggiolo) wrote…</i></summary><blockquote>

Unrelated: this blank line hurts my eyes
</blockquote></details>

Removed.

---

*[cms/server/contest/handlers/base.py, line 85 at r1](https://beta.reviewable.io/reviews/cms-dev/cms/844#-L5dzEGG46aIJyIM4G2M:-L5ezGbT3G4Sj351R5N_:bzaivii) ([raw file](https://github.com/cms-dev/cms/blob/e9e2b62d830babc2e884c7a889b8a75ccbdb07e7/cms/server/contest/handlers/base.py#L85)):*
<details><summary><i>Previously, stefano-maggiolo (Stefano Maggiolo) wrote…</i></summary><blockquote>

Remove entirely?
</blockquote></details>

Will do before pushing, as a later commit adds something back here.

---

*[cms/server/contest/handlers/base.py, line 92 at r1](https://beta.reviewable.io/reviews/cms-dev/cms/844#-L5dzPqs4fG933iffsO7:-L5ezjwF2pmaGXLvc01D:b-p4q5np) ([raw file](https://github.com/cms-dev/cms/blob/e9e2b62d830babc2e884c7a889b8a75ccbdb07e7/cms/server/contest/handlers/base.py#L92)):*
<details><summary><i>Previously, stefano-maggiolo (Stefano Maggiolo) wrote…</i></summary><blockquote>

You missed contest_list
</blockquote></details>

Will do before pushing.

---

*[cms/server/contest/handlers/base.py, line 64 at r2](https://beta.reviewable.io/reviews/cms-dev/cms/844#-L5ejBE94AtRxNlCyEht:-L5ez5BG8s6yCdNgmrP-:b-896fix) ([raw file](https://github.com/cms-dev/cms/blob/1ee07e0fdecf2948722282f3ad7c9e9d3ef63714/cms/server/contest/handlers/base.py#L64)):*
<details><summary><i>Previously, stefano-maggiolo (Stefano Maggiolo) wrote…</i></summary><blockquote>

Comments on what *_translation represent?
</blockquote></details>

Done.

---

*[cms/server/contest/handlers/contest.py, line 333 at r1](https://beta.reviewable.io/reviews/cms-dev/cms/844#-L5e--oKBWU4fSOoVFM5:-L5f0TukDL1IHDz3xcNs:b-p4q5np) ([raw file](https://github.com/cms-dev/cms/blob/e9e2b62d830babc2e884c7a889b8a75ccbdb07e7/cms/server/contest/handlers/contest.py#L333)):*
<details><summary><i>Previously, stefano-maggiolo (Stefano Maggiolo) wrote…</i></summary><blockquote>

Wrap
</blockquote></details>

Will do before pushing.

---

*[cms/server/contest/handlers/contest.py, line 413 at r1](https://beta.reviewable.io/reviews/cms-dev/cms/844#-L5e-6LgEbg8pKKQFokk:-L5f-uiYArbaqLfbMA5u:b-q64pkt) ([raw file](https://github.com/cms-dev/cms/blob/e9e2b62d830babc2e884c7a889b8a75ccbdb07e7/cms/server/contest/handlers/contest.py#L413)):*
<details><summary><i>Previously, stefano-maggiolo (Stefano Maggiolo) wrote…</i></summary><blockquote>

I would have preferred the change in render_params()`s return value to be in a separate commit
</blockquote></details>

See above.

---

*[cms/server/contest/handlers/contest.py, line 320 at r2](https://beta.reviewable.io/reviews/cms-dev/cms/844#-L5ek1RV61waxK1XCXaL:-L5f-5ho4mnR4HEJAf_u:b-abm1j8) ([raw file](https://github.com/cms-dev/cms/blob/1ee07e0fdecf2948722282f3ad7c9e9d3ef63714/cms/server/contest/handlers/contest.py#L320)):*
<details><summary><i>Previously, stefano-maggiolo (Stefano Maggiolo) wrote…</i></summary><blockquote>

What if no avail translations?
</blockquote></details>

available_translations is always non-empty: it is non-empty by default (as it contains DEFAULT_TRANSLATION) and if by filtering it were to become empty we again add back DEFAULT_TRANSLATION

---

*[cms/server/contest/handlers/contest.py, line 326 at r2](https://beta.reviewable.io/reviews/cms-dev/cms/844#-L5ek6Fp6gQKH0YU5epA:-L5f-JOsAzMPnmHhAui5:b6zrrkh) ([raw file](https://github.com/cms-dev/cms/blob/1ee07e0fdecf2948722282f3ad7c9e9d3ef63714/cms/server/contest/handlers/contest.py#L326)):*
<details><summary><i>Previously, stefano-maggiolo (Stefano Maggiolo) wrote…</i></summary><blockquote>

What if cookie_lang is not available?
</blockquote></details>

Good catch.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cms-dev/cms/844)*
<!-- Sent from Reviewable.io -->

lerks commented:
Done, PTAL. Had to force-push. Reviewable may be a bit messed up.

stefano-maggiolo commented on cms/server/contest/handlers/base.py:
```diff
@@ -58,11 +59,25 @@ class BaseHandler(CommonRequestHandler):
+        self.contest_list = {}
+        if self.is_multi_contest():
+            for contest in self.sql_session.query(Contest).all():
+                self.contest_list[contest.name] = contest
```
I can do it later, since it was already like this: this is only needed in the get of this handler, not for all handlers.

lerks replied:
I agree that there's no reason to get *all* contests in *every* request, since they are needed only when showing the contest list. Though be aware that `ContestHandler.choose_contest` uses this dict to pick the contest.

stefano-maggiolo commented on cms/server/contest/handlers/base.py:
```diff
@@ -109,7 +116,7 @@ def write_error(self, status_code, **kwargs):
         # information. If r_params is not defined (i.e. something went
         # *really* bad) we simply return a basic textual error notice.
-        if getattr(self, 'r_params', None) is not None:
+        if self.r_params is not None:
```
Revert or add `self.r_params = None` in `__init__`

lerks replied:
`self.r_params = None` is in the constructor of CommonRequestHandler

stefano-maggiolo commented on cms/server/contest/handlers/contest.py:
```diff
@@ -298,17 +310,11 @@ def _get_current_user_from_cookie(self):
+        lang_codes = list(iterkeys(self.available_translations))
 
         # Select the one the user likes most.
         basic_lang = 'en'
```
DEFAULT_TRANSLATION.identifier?

lerks replied:
Will fix when I push since this is overwritten by a later commit.

stefano-maggiolo commented on cms/server/contest/handlers/contest.py:
```diff
@@ -324,7 +330,8 @@ def setup_locale(self):
         self.set_header("Content-Language", lang_code)
-        self.translation = self.all_translations[lang_code.replace("-", "_")]
+        self.translation = \
+            self.available_translations[lang_code.replace("-", "_")]
```
Nit/ or in another commit - if cookie lang is in lang_codes, we could save looking at headers and parsing and matching.

stefano-maggiolo replied:
Nevermind, I see that automated is used in the rendering regardless.

stefano-maggiolo commented on cms/server/util.py:
```diff
@@ -412,8 +412,8 @@ class CommonRequestHandler(RequestHandler):
-        self.timestamp = None
-        self.sql_session = None
+        self.timestamp = make_datetime()
+        self.sql_session = Session()
```
So you decided to remove expire_all()? I read a bit and it looks safe to remove.

lerks replied:
Yes, I checked too and couldn't find a reason to keep it. Thanks for having double checked.

stefano-maggiolo commented on cms/locale/locale.py:
```diff
@@ -307,41 +307,69 @@ def get_translations():
 
 
-def filter_language_codes(lang_codes, prefix_filter):
+def filter_language_codes(lang_codes, prefixes):
```
Tests for this?

lerks replied:
Will do in later commit.

stefano-maggiolo commented on cms/server/contest/templates/contest.html:
```diff
@@ -59,12 +59,12 @@
-                            <option value=""{% if cookie_lang is None %} selected{% end %}>{{ _("Automatic (%s)") % lang_names[browser_lang] }}</option>
-                        {% for lang_code, lang_name in sorted(iteritems(lang_names), key=lambda x: x[1]) %}
-                            <option value="{{ lang_code }}"{% if cookie_lang == lang_code %} selected{% end %}>{{ lang_name }}</option>
+                            <option value=""{% if cookie_translation is None %} selected{% end %}>{{ _("Automatic (%s)") % automatic_translation.name }}</option>
```
This "automatic" sounds quite odd - maybe default?

lerks replied:
I called it automatic because we render it as "Automatic (language)". Are you suggesting to change both occurrences? I'm open to that but I don't like default as then it conflicts with DEFAULT_TRANSLATION.

stefano-maggiolo commented on cms/server/contest/handlers/contest.py:
```diff
@@ -312,27 +312,30 @@ def _get_current_user_from_cookie(self):
-
-        if self.cookie_lang in http_langs:
-            lang_code = self.cookie_lang
+            LanguageAccept).values()
```
Does it make sense to cache this? parse_accept_header is a pretty complex regex...

stefano-maggiolo commented:
LGTM, possible follow ups not really related to this:
- find a better name than Automatic for the contestant UI;
- load all contests only when necessary;
- cache header to language?

lerks commented:
Merged.
